A15-binary-logging

```diff
@@ -0,0 +1,264 @@
+  // Enumerates the type of logs
+  enum Type {
+    SEND_INITIAL_METADATA = 0;
+    SEND_TRAILING_METADATA = 1;
+    SEND_MESSAGE = 2;
+    RECV_INITIAL_METADATA = 3;
+    RECV_TRAILING_METADATA = 4;
+    RECV_MESSAGE = 5;
+  };
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    CLIENT = 0;
+    SERVER = 1;
+  }
+
+  Type type = 1;      // One of the above Type enum
+  Logger logger = 2;  // One of the above Logger enum
+
+  // Uniquely identifies a call. Each call may have several log entries, they
+  // will share the same call_id. 128 bits split into 2 64-bit parts.
+  Uint128 call_id = 3;
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    // Used by CLIENT_INIT_METADATA, SERVER_INIT_METADATA and TRAILING_METADATA
+    Metadata metadata = 4;
+    // Used by REQUEST and RESPONSE
+    Message message = 5;
+  }
+
+  // Peer address information, will only be recorded in SEND_INITIAL_METADATA
+  // and RECV_INITIAL_METADATA entries.
+  Peer peer = 6;
+};
+
+// Message payload, used by REQUEST and RESPONSE
+message Message {
+  // This flag is currently used to indicate whether the payload is compressed,
+  // it may contain other semantics in the future. Value of 1 indicates that the
+  // binary octet sequence of Message is compressed using the mechanism declared
+  // by the Message-Encoding header. A value of 0 indicates that no encoding of
+  // Message bytes has occurred.
+  uint8 flags = 1;
+  // Length of the message. It may not be the same as the length of the
+  // data field, as the logging payload can be truncated or omitted.
+  uint32 length = 1;
+  // Binary payload, post compression (depends on the flags field), may be
+  // truncated or omitted.
+  bytes data = 2;
+}
+
+// A list of metadata pairs, used in the payload of CLIENT_INIT_METADATA,
+// SERVER_INIT_METADATA and TRAILING_METADATA
+message Metadata {
+  repeated MetadataEntry entry = 1;
+}
+
+// A metadata key value pair
+message MetadataEntry {
+  string key = 1;
+  bytes value = 2;
+}
+
+// Peer information
+message Peer {
+  enum PeerType {
+    PEER_IPV4 = 0;  // peer is struct sockaddr_in
+    PEER_IPV6 = 1;  // peer is struct sockaddr_in6
+    PEER_UNIX = 2;  // peer is struct sockaddr_un
+  };
+  PeerType peer_type = 1;
+  bytes peer = 2;  // value depends on peer_type
+}
+
+// Used to record call_id.
+message Uint128 {
+  fixed64 high = 1;
+  fixed64 low = 2;
+};
+```
+
+A call id should be enclosed in each log record, so that the analyzer can use it
+to group records generated by the same call. The generation of call id can be
+separated from the log generation code, and provided as an interface. In the
+open source version, we can generate and assign call id according to the channel
+and stream information. At the beginning of each RPC, the gRPC client generates
+a call id and sends it to the server in the client initial metadata. This call
```

Comment:
What metadata field will be used for this?

Or should this be removed?  (Have plans changed?)

Reply:
In the current state of things, we *should* log the census span as the call id, and have the option to fall back to an alternate method. I would support simplifying this further and just require that call ids be a monotonically increasing integer. If we do that, we can get rid of `Uint128`.